solvers/fista_automatic_restart.py

```diff
@@ -28,63 +24,66 @@ def set_objective(self, X, y, lmbd, fit_intercept):
+        # more details).
+        inner_iteration = 0
+        n_restarts = 0  # Restart counter
+        last_F = 0
+        ite_per_restart = np.array([int(2*C)])
+        F_tab = np.array([self.cost_function(w)])
+        cb = callback(w)
         # First call of FISTA
-        while i_glob < n_iter and i_int < ite_per_restart[counter]:
+        while cb and inner_iteration < ite_per_restart[n_restarts]:
             w_old = w.copy()
             z -= self.X.T @ (self.X @ z - self.y) / L
             w = self.st(z, self.lmbd / L)
-            z = w + i_int/(i_int+3) * (w - w_old)
-            i_int += 1
-            i_glob += 1
-        counter += 1
-        ite_per_restart[counter] = int(2*C)
-        F_tab[counter] = (.5*np.linalg.norm(self.X @ w - self.y, ord=2)**2
-                          + self.lmbd*abs(w).sum())
-        while i_glob < n_iter:
-            i_int = 0
+            z = w + inner_iteration/(inner_iteration+3) * (w - w_old)
+            inner_iteration += 1
+            cb = callback(w)
+        n_restarts += 1
+        ite_per_restart = np.r_[ite_per_restart, int(2*C)]
+        F_tab = np.r_[F_tab, self.cost_function(w)]
+        while cb:
+            inner_iteration = 0
             # Restart of FISTA after k_tab[counter] iterations
-            while i_glob < n_iter and i_int < ite_per_restart[counter]:
+            while cb and inner_iteration < ite_per_restart[n_restarts]:
                 w_old = w.copy()
                 z -= self.X.T @ (self.X @ z - self.y) / L
                 w = self.st(z, self.lmbd / L)
-                z = w + i_int/(i_int+3) * (w - w_old)
-                i_int += 1
-                i_glob += 1
-            counter += 1
-            if i_glob < n_iter:
-                LastF = (.5 * np.linalg.norm(self.X @ w - self.y, ord=2) ** 2
-                         + self.lmbd * abs(w).sum())
-                F_tab[counter] = LastF
-                # Estimation of the growth parameter
-                mu = (np.min(4 * L / (ite_per_restart[:counter-1]+1) ** 2
-                             * (F_tab[:counter-1] - LastF) /
-                             (F_tab[1:counter] - LastF)))
-                # Update of the number of iterations before next restart
-                if ite_per_restart[counter-1] <= C*np.sqrt(L/mu):
-                    ite_per_restart[counter] = 2 * ite_per_restart[counter-1]
-                else:
-                    ite_per_restart[counter] = ite_per_restart[counter-1]
+                z = w + inner_iteration/(inner_iteration+3) * (w - w_old)
+                inner_iteration += 1
+                cb = callback(w)
+            n_restarts += 1
+            last_F = self.cost_function(w)
+            F_tab = np.r_[F_tab, last_F]
+            # Estimation of the growth parameter
+            mu = np.min(4 * L / (ite_per_restart[:n_restarts-1]+1) ** 2
+                        * (F_tab[:n_restarts-1] - last_F) /
+                        (F_tab[1:n_restarts] - last_F))
```

Comment:
You don't need to specify `n_restart`, this is implicit in `numpy`. Also, `a[-1]` gives you the last value of `a`.
```suggestion
            mu = np.min(4 * L / (ite_per_restart[:-1]+1) ** 2
                        * (F_tab[:-1] - last_F) /
                        (F_tab[1:] - last_F))
```